demos/sslecho/README.md

```diff
@@ -24,3 +24,49 @@ The cert.pem and key.pem files included are self signed certificates with the
 "Common Name" of 'localhost'.
 
 Best to create the 'pem' files using an actual hostname.
+
+Encrypted Client Hello (ECH) Variant
+====================================
+
+``echecho.c`` implements the same functionality but demonstrates minimal code
+changes needed to use ECH. The ``echecho`` binary has the same user interface
+discussed above but enables ECH for the connection, based on hard-coded ECH
+configuration data. A real server would load file(s), and a real client would
+acquire an ECHConfigList from the DNS.
+
+All that's required to use ECH is for the server to enable ECH via
+``SSL_CTX_ech_server_enable_buffer()`` and for the client to do
+similarly via ``SSL_CTX_ech_set1_echconfig()``. Both client and
+server check and print out the status of ECH using ``SSL_ech_get_status()``,
+but that's optional.
+
+To run the server:
+
+            $ LD_LIBRARY_PATH=../.. ./echecho s
+
+To run the client:
+
+            $ LD_LIBRARY_PATH=../.. ./echecho c localhost
+
+All going well both server and client will print the ECH status at the
+start of each connection. That looks like:
+
+            ECH worked (status: 1, inner: localhost, outer: example.com)
+
+If the non-ECH demo client (``sslecho``) is used instead the server will
+output:
+
+            ECH failed/not-tried (status: -101, inner: (null), outer: (null))
+
+If the non-ECH demo server (i.e., ``sslecho``) is used, the client will exit
+with an error as ECH was attempted and failed. In a debug build, that looks
+like:
+
+            80EBEE54227F0000:error:0A000163:SSL routines:tls_process_initial_server_flight:ech required:ssl/statem/statem_clnt.c:3274:
+
+A real client would likely fall back to not using ECH, but the above
```

Comment:
Depends on the client. The ECH protocol fallback is the client authenticates the `public_name` and hopefully gets a set of ECH `retry_configs` (as an encrypted extension from the server). The client then drops that initial TLS session. At that point, what happens next depends on the client - with a browser they'll try again with ECH using the `retry_configs`, before eventually falling back to connecting without ECH (some details browser-dependent, and/or have browser config settings and/or are in flux). For curl as a client, just fail and provide the `retry_config` data if in verbose mode, so that the client can try a 2nd command line invocation.

And note that all the above is liable to see some change in future as part of work on an updated "happy eyeballs" scheme that includes consideration of HTTPS/SVCB RRs. ([This](https://datatracker.ietf.org/doc/draft-pauly-v6ops-happy-eyeballs-v3/) is one proposal in that space that looks like a good starting point to me.)

WRT possible downgrade attack, I think the above is ok, based on authenticating the `public_name` (which is in each entry in the `ECHConfigList` that was retrieved from the origin's DNS zone).